src/token/src/ledger/history.rs

```diff
@@ -0,0 +1,326 @@
+use crate::types::TxRecord;
+use candid::types::{Compound, Serializer, Type};
+use ic_cdk::export::candid::{CandidType, Deserialize, Nat};
+use ic_certified_map::{HashTree, RbTree};
+use serde::de::{SeqAccess, Visitor};
+use serde::Deserializer;
+use std::fmt;
+use std::fmt::Formatter;
+
+#[cfg(target_arch = "wasm32")]
+use ic_certified_map::AsHashTree;
+
+const MAX_TREE_SIZE: usize = 100_000;
+
+#[derive(Default, Clone)]
+pub struct History {
+    tree: RbTree<Vec<u8>, Vec<u8>>,
+    tree_size: usize,
+}
+
+impl History {
+    pub fn insert(&mut self, tx: TxRecord) {
+        self.tree
+            .insert(get_key_bytes(&tx.index), get_tx_bytes(&tx));
+        self.tree_size += 1;
+        if self.tree_size > MAX_TREE_SIZE {
+            self.remove_oldest_tx();
+        }
+    }
+
+    fn remove_oldest_tx(&mut self) {
+        let key = self.tree.iter().map(|(key, _)| key).next().unwrap().clone();
```

Comment:
It is fine, today, to call `unwrap` but if this function is ever called before an entry is inserted this will blow up, or if the `MAX_TREE_SIZE` is changed (it would have to be changed to zero for this to blow up)

```suggestion
        let key = self.tree.iter().map(|(key, _)| key).next().expect("this is only called from `self.insert`").clone();
```